lms/djangoapps/email_marketing/tasks.py

```diff
@@ -145,3 +154,262 @@ def _create_sailthru_user_parm(user, profile, new_user, email_config):
+            raise self.retry(countdown=email_config.sailthru_retry_interval,
+                             max_retries=email_config.sailthru_max_retries)
+
+
+def _build_purchase_item(course_id_string, course_url, unit_cost, mode, course_data, course_id):
+    """
+    Build Sailthru purchase item object
+    :return: item
+    """
+
+    # build item description
+    item = {
+        'id': course_id_string + '-' + mode,
+        'url': course_url,
+        'price': unit_cost * 100,
+        'qty': 1,
+    }
+
+    # get title from course info if we don't already have it from Sailthru
+    if 'title' in course_data:
+        item['title'] = course_data['title']
+    else:
+        try:
+            course = get_course_by_id(course_id)
+            item['title'] = course.display_name
+        except Http404:
+            # can't find, just invent title
+            item['title'] = 'Course ' + course_id_string + ' mode: ' + mode
+
+    if 'tags' in course_data:
+        item['tags'] = course_data['tags']
+
+    # add vars to item
+    sailthru_vars = {}
+    if 'vars' in course_data:
+        sailthru_vars = course_data['vars']
+    sailthru_vars['mode'] = mode
+    sailthru_vars['course_run_id'] = course_id_string
+    item['vars'] = sailthru_vars
+
+    # get list of modes for course and add upgrade deadlines for verified modes
+    for mode_entry in CourseMode.modes_for_course(course_id):
+        if mode_entry.expiration_datetime is not None and CourseMode.is_verified_slug(mode_entry.slug):
+            sailthru_vars['upgrade_deadline_%s' % mode_entry.slug] = \
+                mode_entry.expiration_datetime.strftime("%Y-%m-%d")
+
+    return item
+
+
+def _record_purchase(sailthru_client, email, item, incomplete, message_id, options):
+    """
+    Record a purchase in Sailthru
+    :param sailthru_client:
+    :param email:
+    :param item:
+    :param incomplete:
+    :param message_id:
+    :param options:
+    :return: False it retryable error
+    """
+    try:
+        sailthru_response = sailthru_client.purchase(email, [item],
+                                                     incomplete=incomplete, message_id=message_id,
+                                                     options=options)
+
+        if not sailthru_response.is_ok():
+            error = sailthru_response.get_error()
+            log.error("Error attempting to record purchase in Sailthru: %s", error.get_message())
+            return False
+
+    except SailthruClientError as exc:
+        log.error("Exception attempting to record purchase for %s in Sailthru - %s", email, unicode(exc))
+        return False
+
+    return True
+
+
+def _get_course_content(course_url, sailthru_client):
+    """
+    Get course information using the Sailthru content api.
+
+    If there is an error, just return with an empty response.
+    :param course_url:
+    :param sailthru_client:
+    :return: dict with course information
+    """
+    try:
```

Comment:
Ah, I see you're doing that from a worker. That makes it less critical to cache, but it might still be worth it to reduce the number of outbound requests to Sailthru.